include/avl.h

```diff
@@ -37,16 +38,50 @@ avl_pp generate_avl(int *arr, int len);
 avl_pp init_avl(void);
 
 /* Insert a node in AVL tree */
-bool insert_avl_node(avl_pp head, int val);
+bool insert_avl_node(avl_pp head, int key, int data);
+
 
 /* Delete a node from AVL tree */
-bool delete_avl_node(avl_pp head, int val);
+bool delete_avl_node(avl_pp head, int key);
 
 /* Destroy the tree */
 int destroy_avl(avl_pp head);
+// avl will be destroy isn't necessary sincronize it (is better not destroy)
 
 /* Print a tree in preorder */
 int print_avl(avl_p root, avl_p parent);
 
 /* Traverse tree in BFS to find a given value */
-bool search_BFS_avl(avl_pp root, int val, bool stop);
+int search_BFS_avl(avl_pp root, int key, bool stop);
+
+/*=======================================================*/
+/*         Library ThreadSafe APIs start here            */
+/*=======================================================*/
+
+//The struct is a generalization of avl_pp, to include for every tree one semaphore
+typedef struct avlThSafe
+{
+	avl_pp avlRoot;
+	int semId;
+}avl_pp_S;
+
+/* Generate AVL tree from an array of input values */
+avl_pp_S generate_avl_S(int *arr, int len);
+
+/* Initialize an AVL tree */
+avl_pp_S init_avl_S(void);
+
+/* Insert a node in AVL tree */
+bool insert_avl_node_S(avl_pp_S head, int key, int data);
+
+/* Delete a node from AVL tree */
+bool delete_avl_node_S(avl_pp_S head, int key);
+
+/* Destroy the tree */
+/** [][] avl will be destroy isn't necessary sincronize it (is better not destroy) **/
```

Comment:
Please explain what you mean here. I am unable to understand.

Reply:
i wanted to highlight there aren't any need to create destroy_tree thread-safe (i think)

Reply:
ok